src/consolidate_feature_data.py

```diff
@@ -0,0 +1,239 @@
+
+    return feature_data
+
+
+def make_unique_string_key(owner: str, repo_name: str, issue_number: str) -> str:
+    """
+       Creates a unique 3way string key for identifying every unique feature.
+
+       @param owner: The owner of the repository.
+       @param repo_name: The name of the repository.
+       @param issue_number: The number of the issue.
+
+       @return: The unique string key for the feature.
+    """
+
+    string_key = f"{owner}/{repo_name}/{issue_number}"
+
+    return string_key
+
+
+def merge_feature_and_project_data(feature_data: List[dict],
+                                   project_data_dict: Dict[str, dict],
+                                   project_title: str) -> List[dict]:
+    """
+        Merges feature data with additional project information.
+        Every feature identified by unique string key is compared with keys of project data features.
+        If the keys match, additional information from the project data is added to the feature.
+
+        @param feature_data: The feature data to be merged.
+        @param project_data_dict: The project data to be merged.
+        @param project_title: The title of the project.
+
+        @return: Updated feature data with project data as a list of dictionaries.
+    """
+    # Create list to store features
+    modified_features = []
+
+    # For each feature, add feature data from project
+    for feature in feature_data:
+        owner = feature['Owner']
+        repo_name = feature['RepositoryName']
+        feature_number = feature['Number']
+
+        # Create a key for feature with repo name and issue number
+        string_key = make_unique_string_key(owner, repo_name, feature_number)
+
+        # Create a deep copy of a feature
+        modified_feature = deepcopy(feature)
+
+        # Check if key for feature exists also in the project_data_dict
+        if string_key in project_data_dict:
+            for key, value in project_data_dict[string_key].items():
+                if key not in modified_feature:
+                    modified_feature[key] = value
+
+        # Add project title into the feature
+        modified_feature["ProjectTitle"] = project_title
+
+        # Add the modified feature to the list
+        modified_features.append(modified_feature)
+
+    return modified_features
+
+
+def consolidate_features_with_project() -> Tuple[List[dict], Set[str]]:
+    """
+        Consolidates features that have a project attached.
+        Loading project data and creating project issue dictionary with unique string key.
+        Merging feature and project data with additional info.
+
+        @return: A tuple containing a list of consolidated features with a project and a set of used repository names.
+    """
+    # List to store the feature data with project
+    consolidated_features_with_project = []
+    # Set to store the names of repositories that have been used
+    set_of_used_repos = set()
+
+    if os.path.isdir(PROJECT_DIRECTORY):
```

Comment:
this function is quite nested and I feel like almost a bit too long. Consider simplifying it and perhaps even break it down into multiple functions.

Btw, I like this approach to 'flatten' the code a bit - instead of:

```python
if something:
   do_something_complex_and_nested
```

to:

```python
if not something:
   return

do_something_complex_and_nested
```

by following this second approach, your `do_something_complex_and_nested` code block is not nested any more :)

Reply:
This is cool: https://peps.python.org/pep-0020/

`Flat is better than nested.`

Reply:
As this was overtaken by migration, let's make this a note to refactor SPIKE @MobiTikula .